### ISSUE TITLE
Exception: Data, labels and weights should have the same number of rows.

### ISSUE THREAD

**Issue by malakar-soham** [interpretml/interpret#308]
Hey,

I am working on an NLP problem( multilabel, multiclass), using EBM and I'm supplying dataframes of the following shape where X is a feature_vector and y is a multilabel binerized vector to `model.fit(X,y)`:
``` python
X = 2237, 604 (sparse matrix)
y = 2237, 6 (ndarray)
```

The number of rows in the features and labels are same but while fitting the EBM in ebm.py **Line 1003**:
`estimators = provider.parallel(BaseCoreEBM.fit_parallel, train_model_args_iter)`

These are the following shapes:
```python
X: 2237, 604
w: 13422
y: 13422
```

The w and y seems to have been flattened which is why there is an error. But the y supplied is a [multilabel binerized](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html) vector. 

Is there any workaround this ?

**Comment by interpret-ml** [comment]
Hi @malakar-soham --

We support multiclass, but not multilabel.  Unlike DNNs, there probably isn't a large benefit to multitask learning for EBMs, so we'd suggest training separate models for each label.  Having said that, we should instead throw a helpful exception if multi-dimensional data like this is passed in.  Fortunately, the current code in the develop branch does throw such an exception due to a recent change, so this oversight should be corrected in our next release.

-InterpretML team

**Comment by malakar-soham** [comment]
Okay cool

**Comment by RobertKuebler** [comment]
Hello! 

But sklearn supports multi-outputs for tree models as well. Even if the benefits for predictive power are low, being able to train one model for several outputs greatly decreases the run time. That's why I think it would still be great if you could support this. Thanks a lot! :)

Best
Robert

**Comment by OldPatrick** [comment]
Hello, I believe I have to second that:
If I use the adult data from your EBM Example:
<img width="812" alt="image" src="https://user-images.githubusercontent.com/75302135/171143791-981c56d6-f4fa-49e7-8252-dc094e748868.png">

and use "CapitalGain" and "HoursPerWeek"  as target"s".
It works fine and well with a DecisionTreeRegressor:

<img width="562" alt="image" src="https://user-images.githubusercontent.com/75302135/171144203-b3a63af2-4e55-481f-acd7-94ed8a842254.png">

**Comment by interpret-ml** [comment]
Hi @RobertKuebler & @OldPatrick --

In terms of optimizations to make building EBMs faster, multi-output probably doesn't have the same benefit that it might have in other tree building algorithms.  Other tree algorithms tend to spend most of their time figuring out where to put splits, and once those splits are determined they need to separate the data into left and right sides.  The EBM building process tends to spend most of its time generating histograms on the entirety of the dataset and updating residuals.  All our splits happen within a single feature, which allows us to avoid the work separating data once a split is chosen.

One thing I've noticed is that XGBoost and LightGBM build separate trees per-class in multi-class.  From the code it looks like they spent extra effort doing this.  My speculative guess is that boosted trees get more benefit by allowing different splits per-class than the benefit they get from sharing information when choosing split locations, especially when there are minority classes where some splits should be disallowed due to lack of samples.  One surprising observation we've made is that the precise location of splits doesn't seem to affect the accuracy of EBM models much.  In differentially privacy EBMs we choose completely random split locations, and the cost of doing this seem to be minor, so that's one indication that there's not much benefit left on the table that multi-output can exploit.

I do note though that XGBoost recently added support for multi-output regression using custom objectives, so that's one place where multi-output might potentially have some benefit.

-InterpretML team
